Hook up shipping methods to API

Aljullu opened the pull request:
This PR hooks up the shipping methods list in the _Cart_ and _Checkout_ components with the API endpoints introduced in #1344.

### Screenshots

![Peek 2020-01-19 12-32](https://user-images.githubusercontent.com/3616980/72680232-cebadb80-3ab7-11ea-96d4-a6d297a656b7.gif)

### How to test the changes in this Pull Request:

1. Make sure you have different shipping methods set up and create a post with a _Checkout_ block.
2. Fill the checkout _Shipping_ step and verify shipping rates update accordingly.
3. Test again setting some package rules using the [WooCommerce Advanced Shipping Packages](https://woocommerce.com/products/woocommerce-advanced-shipping-packages/#) extension.

### Follow-ups
I left some `@todo` comments in the code with tasks to be fixed in the future and I created issues for them:
* Radio inputs are a bit misaligned, but I think it's better to fix that when we update its styling to match `@wordpress/components` (#1556).
* Display product name and quantity before package options (#1554).
* Add UI for loading state of shipping rates (#1555).
* Display currently selected location in Cart no results message (#1606).

### To do in this PR
* [x] Rebase & fix conflicts.
* [x] Add debounce (d22af0be78ea4ce64e534c524fdc5301409ddc6c).
* [x] Make endpoint return an error if country is invalid (602141d900caecd6234d3da80b9bbfa2aca1f601).
* [x] Don't display radio input when there is only one option (c6130f09033bbfb4d8797f43cc2351616196aca9).

nerrad commented:
@Aljullu is this pull up for review? I ask because although it's in the QA/Review zenhub pipeline it's still a draft pull and has an in progress label.  I'm happy to review but didn't want to start on that if there's still work to be done on it before that.

Aljullu commented:
> @Aljullu is this pull up for review?

Not yet, I want to have #1411 merged into `master` first because I expect some rebase conflicts when that happens. Even though #1411 is approved, I would like to investigate how can we decrease the size of the build files before merging.

> I ask because although it's in the QA/Review zenhub pipeline it's still a draft pull and has an in progress label.

Oh, that was a mistake. I moved this PR back to the _In Progress_ pipeline.

garymurray commented:
> When there are several packages, we need to display the list of products that will be sent with each package. Currently it's shown this way:

@Aljullu are you looking for how the cart/checkout appears if it has for example two products with two different shipping methods for each item?

Aljullu commented:
> @Aljullu are you looking for how the cart/checkout appears if it has for example two products with two different shipping methods for each item?

@garymurray Yes, exactly. I guess we will want to list the products in each package and its quantity, like it's shown in the screenshot above of the current Cart.

garymurray commented:
Commented and added some design ideas here for feedback: https://github.com/woocommerce/woocommerce-gutenberg-products-block/issues/1554

nerrad commented:
Before I get too deep into a review, I did some testing and this isn't working for me.  First here's a zone I setup for testing with:

<img width="1717" alt="Image 2020-01-16 at 3 00 53 PM" src="https://user-images.githubusercontent.com/1429108/72558610-097f0280-3871-11ea-89e2-c3105389f547.png">

The price value for the flat rate in this zone is $5.00.

In my testing, I'm not seeing any expected shipping options when I fill in the values that I _expect_ to see options for.  Here's some more issues:

- When I type in the country, the City is automatically field is filled in with the same value (appears to happen in your screenshot too?) but there is no city included with the request. Here's my query string parameters for the request:

```
address_1: 123 Anytown St.
state: Ontario
postcode: N2T2Y4
country: Canada
```

- This is mostly display, but it's not obvious that I put my province in the _county_ field. This field is geographically specific so I would expect it to update when the country is selected. Is this something that has a followup already?

The pressing issue for me is that I'm not seeing expected results for the above value, this _might_ be an issue with the shipping endpoint though because this is the response I get for the above request:

```json
[
  {
    "destination": {
      "address_1": "123 Anytown St.",
      "address_2": null,
      "city": null,
      "state": "Ontario",
      "postcode": "N2T2Y4",
      "country": "CANADA"
    },
    "items": [
      "6f4922f45568161a8cdf4ad2299f6d23",
      "9bf31c7ff062936a96d3c8bd1f8f2ff3",
      "aab3238922bcc25a6f606eb525ffdc56"
    ],
    "shipping_rates": []
  }
]
```

Here's the actual generated request url:

```
https://woo.test/wp-json/wc/store/cart/shipping-rates?address_1=123%20Anytown%20St.&state=Ontario&postcode=N2T2Y4&country=Canada&_locale=user
```

nerrad commented:
Also should note that (without having looked at the code yet), this appears it could use some debouncing. I'm getting a request on every key press when in the country/province fields and it seems kind of excessive.

Aljullu commented:
> When I type in the country, the City is automatically field is filled in with the same value (appears to happen in your screenshot too?) but there is no city included with the request. Here's my query string parameters for the request:

Right, that was fixed in #1580. Will rebase this PR so it's fixed here too.

> This is mostly display, but it's not obvious that I put my province in the county field. This field is geographically specific so I would expect it to update when the country is selected. Is this something that has a followup already?

Agree with you. I was thinking to take a look at it as part of #1575. Added a specific bullet point in the breakdown so we don't forget. :+1: 

> The pressing issue for me is that I'm not seeing expected results for the above value, this might be an issue with the shipping endpoint though because this is the response I get for the above request:

I think that might be because you are passing `CANADA` as the country instead of `CA`. I will update the endpoint so it returns an error when the `country` is not valid. So this kind of issues are easier to debug in the future.

In any case, having the `<CountryInput>` component will make this not to happen to users.

> Also should note that (without having looked at the code yet), this appears it could use some debouncing. I'm getting a request on every key press when in the country/province fields and it seems kind of excessive.

Sure, that makes sense! I added a `To do in this PR` section in the leading comment with the list of tasks and moved this PR back to `In Progress`.

Aljullu commented:
I rebased and addressed the feedback left in this PR. It should be ready for another look.

nerrad commented on src/RestApi/StoreApi/Controllers/CartShippingRates.php:
```diff
@@ -82,8 +82,24 @@ public function get_items( $request ) {
+					),
+					[ 'status' => 400 ]
+				);
+			}
```
I like this additional context added to this error 👍

Aljullu commented on assets/js/base/components/radio-control/index.js:
```diff
@@ -40,3 +40,5 @@ const RadioControl = ( {
 };
 
 export default withComponentId( RadioControl );
+export { default as RadioControlOption } from './option';
```
Good catch! Done in f606fef9ca4204017da63ac0ebaec488a5321723.

Aljullu commented on assets/js/base/components/shipping-rates-control/index.js:
```diff
@@ -0,0 +1,146 @@
+		);
+	};
+
+	const renderSeveralOptions = ( shippingRates, i ) => {
```
Yes, I wasn't very happy with these functions' signature... I completely refactored this in aaf2cb1c4c3eedd19cf7b7bb273755f59b5b2995. Basically I split it into several components so there is no longer the need to pass `i` around.

Aljullu commented:
Thanks for the review @nerrad! I fixed the conflicts in this branch and applied your feedback. This should be ready for another review.

> (Checkout Block) The content jumps while the new shipping rates are being retrieved. We should try to reduce this jumpiness by only swapping in new methods after we have them. I'm guessing this is to be handled by #1555 ?

Right, the plan is to improve this in #1555. We will not be able to reduce jumpiness completely because depending on the address there might be more or less options, but yes, currently when it's loading the block completely disappears and this could be improved adding placeholders or graying out the previous options.

> (Cart Block) - I'm unable to pull up the shipping address on the frontend (don't think that's wired up yet). But the shipping options without zone information seems correct.

Yes, currently the Cart block doesn't pass any address to the backend, so it just shows the shipping rates for your default location.

mikejolley commented on src/RestApi/StoreApi/Controllers/CartShippingRates.php:
```diff
@@ -246,6 +262,9 @@ protected function get_shipping_packages( $request ) {
 		$packages = WC()->cart->get_shipping_packages();
 
 		foreach ( $packages as $key => $package ) {
+			if ( ! $request['country'] ) {
```
Why is this within the loop? If the country is not set, the loop over $packages won't be needed.

Aljullu replied:
Oh, right. That didn't make a lot of sense. :sweat_smile: 

Good catch, fixed in ac0f834f5e843ae035dfef8206f0b4eac4d86dfd.

haszari commented:
Noticed an issue with html entities in shipping method name when testing #1741. 

https://github.com/woocommerce/woocommerce-gutenberg-products-block/issues/1759
